test_pytest/test_bezier_extraction.py

```diff
@@ -0,0 +1,118 @@
+        )
+
+
+@pytest.mark.parametrize("splinetype", all_splinetypes)
+def test_extraction_matrices(splinetype, request):
+    spline = request.getfixturevalue(splinetype)
+
+    if splinetype == "bspline_3p3d":
+        """Create matrices to extract splines"""
+        # Init b-splines
+        bspline = spline
+        bspline.elevate_degrees([0, 1, 2])
+        bspline.insert_knots(0, np.random.rand(3))
+        bspline.insert_knots(1, np.random.rand(3))
+        bspline.insert_knots(2, np.random.rand(3))
+
+        # Extract splines
+        # BSpline
+        beziers_b = bspline.extract_bezier_patches()
+        b_matrices = bspline.knot_insertion_matrix(beziers=True)
+        for m, b in zip(b_matrices, beziers_b):
+            # Test matrices m against spline ctps
+            assert np.allclose(b.control_points, m @ bspline.control_points)
+
+    elif splinetype == "nurbs_3p3d":
+        """Create matrices to extract splines"""
+        # Init nurbs
+        nurbs = spline
+        nurbs.elevate_degrees([0, 1, 2])
+        nurbs.insert_knots(0, np.random.rand(3))
+        nurbs.insert_knots(1, np.random.rand(3))
+        nurbs.insert_knots(2, np.random.rand(3))
+        # NURBS
+        n_matrices = nurbs.knot_insertion_matrix(beziers=True)
+        beziers_n = nurbs.extract_bezier_patches()
+        for m, b in zip(n_matrices, beziers_n):
+            # Test matrices m against spline ctps
+            assert np.allclose(b.weights, m @ nurbs.weights)
+            assert np.allclose(
+                b.control_points,
+                (m @ (nurbs.control_points * nurbs.weights)) / b.weights,
+            )
+
+    elif splinetype == "bspline_2p2d":
+        """Create matrices to extract splines"""
+        # Init nurbs
+        bspline = spline
+        bspline.elevate_degrees([0, 1])
+        bspline.insert_knots(0, np.random.rand(3))
+        bspline.insert_knots(1, np.random.rand(3))
+        # NURBS
+        n_matrices = bspline.knot_insertion_matrix(beziers=True)
+        beziers_n = bspline.extract_bezier_patches()
+        for m, b in zip(n_matrices, beziers_n):
+            # Test matrices m against spline ctps
+            assert np.allclose(b.control_points, m @ bspline.control_points)
+
+    elif splinetype == "nurbs_2p2d":
+        """Create matrices to extract splines"""
+        # Init nurbs
+        nurbs = spline
+        nurbs.elevate_degrees([0, 1])
+        nurbs.insert_knots(0, np.random.rand(3))
+        nurbs.insert_knots(1, np.random.rand(3))
+        # NURBS
+        n_matrices = nurbs.knot_insertion_matrix(beziers=True)
+        beziers_n = nurbs.extract_bezier_patches()
+        for m, b in zip(n_matrices, beziers_n):
+            # Test matrices m against spline ctps
+            assert np.allclose(b.weights, m @ nurbs.weights)
+            assert np.allclose(
+                b.control_points,
+                (m @ (nurbs.control_points * nurbs.weights)) / b.weights,
```

Comment:
```suggestion
    if "3d" in splinetype:
        spline.elevate_degrees([0, 1, 2])
        spline.insert_knots(0, np.random.rand(3))
        spline.insert_knots(1, np.random.rand(3))
        spline.insert_knots(2, np.random.rand(3))
    elif "2d" in splinetype:
        spline.elevate_degrees([0, 1])
        spline.insert_knots(0, np.random.rand(3))
        spline.insert_knots(1, np.random.rand(3))
    else:
        ytest.fail()
    n_matrices = spline.knot_insertion_matrix(beziers=True)
    beziers_n = spline.extract_bezier_patches()
    for m, b in zip(n_matrices, beziers_n):
        # Test matrices m against spline ctps
        if "nurbs" in splinetype:
            assert np.allclose(b.weights, m @ spline.weights)
            assert np.allclose(
                b.control_points,
                (m @ (spline.control_points * spline.weights)) / b.weights,
            )
        elif "bspline" in splinetype:
            assert np.allclose(b.control_points, m @ spline.control_points)
        else:
            pytest.fail()
```
I think you can make this more compact and not reuse the same code over and over again. Please look over this if this is actually correct and test it, I have used the online editor to make this edit so mistakes are likely.

If I oversaw something that makes this compressed implementation not possible please check if you can compress this in another way. Or split it up in multiple functions since no code is reused between the if clauses anyway and it makes it more readable split up in that case.